compiler/rustc_query_impl/src/lib.rs

```diff
@@ -17,37 +18,217 @@ extern crate rustc_macros;
+use rustc_middle::query::erase::{erase, restore, Erase};
 use rustc_middle::query::Key;
+use rustc_middle::ty::query::QueryCaches;
+use rustc_middle::ty::query::QueryEngine;
 use rustc_middle::ty::query::{
     query_keys, query_provided, query_provided_to_value, query_storage, query_values,
 };
-use rustc_middle::ty::query::{ExternProviders, Providers, QueryEngine};
 use rustc_middle::ty::TyCtxt;
+use rustc_query_system::dep_graph::DepNodeParams;
+use rustc_query_system::ich::StableHashingContext;
+use rustc_query_system::Value;
 use rustc_span::Span;
+use std::marker::PhantomData;
 
 #[macro_use]
 mod plumbing;
-pub use plumbing::QueryCtxt;
+pub use plumbing::{Queries, QueryCtxt};
 use rustc_query_system::query::*;
 #[cfg(parallel_compiler)]
 pub use rustc_query_system::query::{deadlock, QueryContext};
 
 pub use rustc_query_system::query::QueryConfig;
+use rustc_query_system::HandleCycleError;
 
 mod on_disk_cache;
 pub use on_disk_cache::OnDiskCache;
 
 mod profiling_support;
 pub use self::profiling_support::alloc_self_profile_query_strings;
 
-rustc_query_append! { define_queries! }
+trait Bool {
+    fn value() -> bool;
+}
+
+struct True;
+
+impl Bool for True {
+    fn value() -> bool {
+        true
+    }
+}
+struct False;
+
+impl Bool for False {
+    fn value() -> bool {
+        false
+    }
+}
+
+#[derive(Copy, Clone)]
+struct DynamicQueryStatic {
+    name: &'static str,
+    eval_always: bool,
+    dep_kind: rustc_middle::dep_graph::DepKind,
+    handle_cycle_error: HandleCycleError,
+}
+
+struct DynamicQuery<'tcx, C: QueryCache> {
+    base: DynamicQueryStatic,
+    query_state: FieldOffset<QueryStates<'tcx>, QueryState<C::Key, crate::dep_graph::DepKind>>,
+    query_cache: FieldOffset<QueryCaches<'tcx>, C>,
+    cache_on_disk: fn(tcx: TyCtxt<'tcx>, key: &C::Key) -> bool,
+    execute_query: fn(tcx: TyCtxt<'tcx>, k: C::Key) -> C::Value,
+    compute: fn(tcx: TyCtxt<'tcx>, key: C::Key) -> C::Value,
+    try_load_from_disk:
+        fn(qcx: QueryCtxt<'tcx>, idx: &C::Key) -> TryLoadFromDisk<QueryCtxt<'tcx>, C::Value>,
+    hash_result: Option<fn(&mut StableHashingContext<'_>, &C::Value) -> Fingerprint>,
+    from_cycle_error:
+        fn(tcx: TyCtxt<'tcx>, cycle: &[QueryInfo<crate::dep_graph::DepKind>]) -> C::Value,
+}
+
+struct DynamicConfig<'tcx, C: QueryCache, Anon, DepthLimit, Feedable> {
+    dynamic: &'tcx DynamicQuery<'tcx, C>,
+    data: PhantomData<(Anon, DepthLimit, Feedable)>,
+}
+
+impl<'tcx, C: QueryCache, Anon, DepthLimit, Feedable> Copy
+    for DynamicConfig<'tcx, C, Anon, DepthLimit, Feedable>
+{
+}
+impl<'tcx, C: QueryCache, Anon, DepthLimit, Feedable> Clone
+    for DynamicConfig<'tcx, C, Anon, DepthLimit, Feedable>
+{
+    fn clone(&self) -> Self {
+        DynamicConfig { dynamic: self.dynamic, data: PhantomData }
+    }
+}
+
+impl<'tcx, C: QueryCache, Anon: Bool, DepthLimit: Bool, Feedable: Bool> QueryConfig<QueryCtxt<'tcx>>
+    for DynamicConfig<'tcx, C, Anon, DepthLimit, Feedable>
```

Comment:
IIUC, we now have 2 (Anon) x 2 (DepthLimit) x 2 (Feedable) x 3 (types of cache) x (number of value sizes) instance of the functions in `rustc_query_system`.
Is there much gained by having anon/depth_limit/feedable statically, instead of plain booleans?

Reply:
Anon, DepthLimit and Feedable have few instances, so the cost of specializing for them is quite low at the moment. Fully erased we reduce things to 94 instances so few additional ones isn't impactful.